crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,345 @@
+                                data.iter()
+                                    .flatten()
+                                    .copied()
+                                    .map(|c| Complex64::new(c.re.into(), c.im.into()))
+                                    .collect(),
+                            )
+                            .map_err(|_| {
+                                RegisterMatrixConversionError::InvalidShape {
+                                    register: name.to_string(),
+                                }
+                            })?,
+                        ),
+                    };
+                    Ok((name.clone(), register_matrix))
+                })
+                .collect::<Result<HashMap<String, RegisterMatrix>, RegisterMatrixConversionError>>(
+                )?,
+        ))
     }
 
-    /// `readout_values` maps program-defined readout to result-defined readout, e.g.:
-    ///     { "ro[0]": "q0", "ro[1]": "q1" }
-    /// where `ro[0]` is defined in the original program, and `q0` is what comes back in execution results.
-    /// Here we map the result-defined readout values back to their original result-defined names.
-    pub(crate) fn from_mappings_and_values(
-        readout_mappings: &HashMap<String, String>,
-        readout_values: &HashMap<String, ReadoutValues>,
-    ) -> Self {
-        let result = readout_values
-            .iter()
-            .flat_map(|(readout_name, values)| {
-                readout_mappings
+    /// Attempts to build a [`ReadoutMap`] from [`QpuReadout`].
+    ///
+    /// # Errors
+    ///
+    /// This fails if the underlying [`QpuReadout`] data is jagged. See [`ReadoutData`] for more
+    /// detailed explanations of why and when this occurs.
+    pub fn from_qpu_readout_data(
+        readout_data: &QpuReadout,
+    ) -> Result<Self, RegisterMatrixConversionError> {
+        Ok(
+            Self(
+                readout_data
+                    .mappings
                     .iter()
-                    .filter(|(_, program_alias)| *program_alias == readout_name)
-                    .map(|(program_name, _)| program_name.as_ref())
-                    .map(parse_readout_register)
-                    .map(|reference| Ok((reference?, values.clone())))
-                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
-            })
-            .flatten()
-            .collect::<HashMap<_, _>>()
-            .into();
-
-        result
-    }
-}
+                    // Pair all the memory references with their readout values
+                    .map(|(memory_reference, alias)| {
+                        Ok((
+                            parse_readout_register(memory_reference).map_err(|e| {
+                                RegisterMatrixConversionError::MemoryReferenceParseError(e)
+                            })?,
+                            readout_data.readout_values.get(alias).ok_or(
```

Comment:
Why not this, to avoid cloning later on?

```suggestion
                            readout_data.readout_values.remove(alias).ok_or(
```

Reply:
We'd have to make `QpuReadout` mutable and consume it to do so right? If so, I don't think we can, since we don't want to destroy any of the data in `QpuReadout` if the transformation could ultimately fail.